tsl/src/compression/compression.c

```diff
@@ -2483,6 +2488,24 @@ decompress_chunk_walker(PlanState *ps, List *relids)
 							 errhint("Set timescaledb.enable_dml_decompression to TRUE.")));
 
 				decompress_batches_for_update_delete(current_chunk, predicates);
+
+				/* This is a workaround specifically for bitmap heap scans:
+				 * during node initialization, initialize the scan state with the active snapshot
+				 * but since we are inserting data to be modified during the same query, they end up
+				 * missing that data by using a snapshot which doesn't account for this decompressed
+				 * data. To circumvent this issue, we change the internal scan state to use the
+				 * transaction snapshot and execute a rescan so the scan state is set correctly and
+				 * includes the new data.
+				 */
+				if (should_rescan)
+				{
+					ScanState *ss = ((ScanState *) ps);
+					if (ss && ss->ss_currentScanDesc)
+					{
+						ss->ss_currentScanDesc->rs_snapshot = GetTransactionSnapshot();
```

Comment:
Postgres almost never uses `GetTransactionSnapshot()` like this. Why aren't we using `CommandCounterIncrement` like e.g. `fmgr_sql` that calls SQL functions that can modify tables?

Reply:
They actually they do use `GetTransactionSnapshot()`: https://github.com/postgres/postgres/blob/master/src/backend/executor/functions.c#L1180

I think this snapshot manipulation can be done better but at this point this is the minimal change to fix the issue.

Reply:
It only does that once if it doesn't have a snapshot already. And also does `PushActiveSnapshot` (no idea what this is). I also wonder how exactly it works in Repeatable Read, when GetSnapshotData always returns the same snapshot, the first in transaction.

Anyway, don't mind me, I realized I'll need to spend a week on this to meaningfully comment :smile: Just saw an unusual thing.

Reply:
> Anyway, don't mind me, I realized I'll need to spend a week on this to meaningfully comment  Just saw an unusual thing.

The only thing I would suggest is to test this under repeatable read. The correct visibility-handling code should work for all isolation levels without much special-casing.

Reply:
Isolation tests with various isolation level combinations will be added in PR #5675 

Verified those tests against this PR produced no issues.